runtime/lua/vscode-neovim/api.lua

```diff
@@ -1,133 +1,205 @@
-local api = vim.api
-local fn = vim.fn
--- used to execute vscode command
-local command_event_name = "vscode-command"
--- used for extension communications
-local plugin_event_name = "vscode-neovim"
-
 local M = {}
 
--- send commands to vscode
-function M.notify(command, ...)
-  return vim.rpcnotify(vim.g.vscode_channel, command_event_name, command, { ... })
-end
-
-function M.call(command, ...)
-  return vim.rpcrequest(vim.g.vscode_channel, command_event_name, command, { ... })
-end
-
--- send commands to vscode extension
-function M.notify_extension(command, ...)
-  return vim.rpcnotify(vim.g.vscode_channel, plugin_event_name, command, { ... })
-end
-
-function M.call_extension(command, ...)
-  return vim.rpcrequest(vim.g.vscode_channel, plugin_event_name, command, { ... })
-end
+------- Requests -------
 
--- send command to vscode with range (line or char). [1, 1]-based.
-function M.call_range(command, line1, line2, leaveSelection, ...)
-  return M.call_extension("range-command", command, "V", line1, line2, 1, 1, leaveSelection, { ... })
-end
+local REQUEST_STATE = {
+  id = 0,
+  callbacks = {},
+}
 
-function M.notify_range(command, line1, line2, leaveSelection, ...)
-  return M.notify_extension("range-command", command, "V", line1, line2, 1, 1, leaveSelection, { ... })
+local function add_callback(callback)
+  REQUEST_STATE.id = REQUEST_STATE.id + 1
+  REQUEST_STATE.callbacks[REQUEST_STATE.id] = callback
+  return REQUEST_STATE.id
 end
 
-function M.call_range_pos(command, line1, line2, pos1, pos2, leaveSelection, ...)
-  return M.call_extension("range-command", command, "v", line1, line2, pos1, pos2, leaveSelection, { ... })
+---Invoke the callback, called by vscode
+---@param id number callback id
+---@param result any resutl
+---@param is_error boolean is error
+function M.invoke_callback(id, result, is_error)
+  vim.schedule(function()
+    local callback = REQUEST_STATE.callbacks[id]
+    REQUEST_STATE.callbacks[id] = nil
+    if callback then
+      if is_error then
+        callback(result, nil)
+      else
+        callback(nil, result)
+      end
+    end
+  end)
 end
 
-function M.notify_range_pos(command, line1, line2, pos1, pos2, leaveSelection, ...)
-  return M.notify_extension("range-command", command, "v", line1, line2, pos1, pos2, leaveSelection, { ... })
+---Run action, this function is asynchronous
+---@param name string The action name, generally a vscode command
+---@param opts? table Optional options table, all fields are optional
+---            - args: (table) Optional arguments for the action
+---            - range: (table) Specific range for the action. In visual mode, this parameter is generally not required.
```

Comment:
Before merge, we (I can) need to document this in the readme, especially an example to specify range. 

Also, I'm not totally sure about the range format. I think lsp range is useful, but in reality I'm not sure how often it will be used. The line_range example shows that. 

My suggestion:

```lua
range = {start: {int | {line: int, character: int}, end: {int | {line: int, character: int}} | [int | {line: int, character: int}, int | {line: int, character: int}]

range = {vim.fn.line("."), vim.fn.line("v")}
range = {{line=vim.fn.line("."), character=vim.fn.col(".")}, {line=vim.fn.line("v"), character=vim.fn.col("v")}}
--- still supports lsp range
```

Reply:
We can provide some convenient functions to create range, in fact, we can use vim.lsp.util directly.

Reply:
Either we provide a flexible API to accept many different types of ranges, or we create external helpers. Not a mix of both. If we want to make helpers, then we should remove the `line_range` as it will be redundant.

What we need, either provided by the API or by helpers:
* lsp range
* convenient/concise char-wise range
* convenient/concise line-wise range
* easy conversion from `vim.fn.getpos(".")` to char or line-wise range

Reply:
> Before merge, we (I can) need to document this in the readme

I will write the wiki so that more people can improve it together. Write examples at the same time

Reply:
> Either we provide a flexible API to accept many different types of ranges, or we create external helpers. Not a mix of both. If we want to make helpers, then we should remove the `line_range` as it will be redundant.
> 
> What we need, either provided by the API or by helpers:
> 
> * lsp range
> * convenient/concise char-wise range
> * convenient/concise line-wise range
> * easy conversion from `vim.fn.getpos(".")` to char or line-wise range

Makes sense

Reply:
Maybe all we need is `range = vscode.util.mark_to_range(".", linewise=true)`, `range = vscode.util.linewise(3, 4)`, `range = vscode.util.charwise({3, 0}, {4, 1})`.

Reply:
> Maybe all we need is `range = vscode.util.mark_to_range(".", linewise=true)`, `range = vscode.util.linewise(3, 4)`, `range = vscode.util.charwise({3, 0}, {4, 1})`.

It seems that these functions are not necessary. Firstly, specifying the range is rarely needed. Secondly, the range parameter is flexible enough, users just need to remember to use 0-indexed values.